core/external_search.py

```diff
@@ -123,369 +124,165 @@ class ExternalSearchIndex(HasSelfTests):
-                # then using the alias, use the index directly instead
-                # of going through the alias.
-                _use_as_works_alias(self.works_index)
-            return
-
-        # Create the alias and search against it.
-        response = self.indices.put_alias(index=self.works_index, name=alias_name)
-        if not response.get("acknowledged"):
-            self.log.error("Alias '%s' could not be created", alias_name)
-            # Work against the index instead of an alias.
-            _use_as_works_alias(self.works_index)
-            return
-        _use_as_works_alias(alias_name)
-
-    def setup_index(self, new_index=None, **index_settings):
-        """Create the search index with appropriate mapping.
-
-        This will destroy the search index, and all works will need
-        to be indexed again. In production, don't use this on an
-        existing index. Use it to create a new index, then change the
-        alias to point to the new index.
-        """
-        index_name = new_index or self.works_index
-        if self.indices.exists(index_name):
-            self.log.info("Deleting index %s", index_name)
-            self.indices.delete(index_name)
-
-        self.log.info("Creating index %s", index_name)
-        body = self.mapping.body()
-        body.setdefault("settings", {}).update(index_settings)
-        index = self.indices.create(index=index_name, body=body)
-
-    def set_stored_scripts(self):
-        for name, definition in self.mapping.stored_scripts():
-            # Make sure the name of the script is scoped and versioned.
-            if not name.startswith("simplified."):
-                name = self.mapping.script_name(name)
-
-            # If only the source code was provided, configure it as a
-            # Painless script.
-            if isinstance(definition, (bytes, str)):
-                definition = dict(script=dict(lang="painless", source=definition))
-
-            # Put it in the database.
-            self.put_script(name, definition)
-
-    def transfer_current_alias(self, _db, new_index):
-        """Force -current alias onto a new index"""
-        if not self.indices.exists(index=new_index):
-            raise ValueError("Index '%s' does not exist on this client." % new_index)
-
-        current_base_name = self.base_index_name(self.works_index)
-        new_base_name = self.base_index_name(new_index)
-
-        if new_base_name != current_base_name:
-            raise ValueError(
-                (
-                    "Index '%s' is not in series with current index '%s'. "
-                    "Confirm the base name (without version number) of both indices"
-                    "is the same."
-                )
-                % (new_index, self.works_index)
+            self._search_write_pointer = self._search_service.write_pointer_name(
+                self._revision_base_name
             )
+            self._search_migrator = SearchMigrator(
```

Comment:
@io7m what is the purpose of having a migrator then cancelling the migration in the init method?

Reply:
@io7m :point_up:

Reply:
Sorry, wasn't getting notifications for some reason. Checking!

Reply:
Creating a migration has side effects. We want the side effects. However, we don't actually have all of the works and search documents we'll need at this point, so we cancel the rest of the migration.

I'm not completely happy with it either, but the structure of the existing CM code meant that there had to be _some_ initialization work done up-front to allow the system to work, with the actual migration (creation and submission of search documents and the like) happening a bit later.

Reply:
Could we have a method -- with a descriptive name -- that just evokes those side effects and handles the `cancel` itself?

Reply:
I think `initialize_versioning()` is the most literal description of what the method would actually be doing.

Reply:
Is the side-effect we're looking for just the creation of the empty index? Because even if we create the versioned index, we are not updating the read and write _pointers_ unless we do a `finish()` so the `ExternalSearchIndex` object cannot use the index anyways.

Reply:
@io7m @tdilauro before merging this in, I would like to resolve this conversation,

Reply:
@io7m :point_up:

Reply:
That's a good point. The effect was supposed to be to get things to `S6` and then stop:

https://ataxia.io7m.com/2023/06/08/search/

So the read pointer should be created if it doesn't exist. If that's not happening, then this is a bug. :sweat: